packages/compiler/src/render3/partial/ng_module.ts

```diff
@@ -0,0 +1,67 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import * as o from '../../output/output_ast';
+import {Identifiers as R3} from '../r3_identifiers';
+import {createNgModuleType, R3NgModuleDef, R3NgModuleMetadata} from '../r3_module_compiler';
+import {refsToArray} from '../util';
+import {DefinitionMap} from '../view/util';
+
+import {R3DeclareNgModuleMetadata} from './api';
+
+
+export function compileDeclareNgModuleFromMetadata(meta: R3NgModuleMetadata): R3NgModuleDef {
+  const definitionMap = createNgModuleDefinitionMap(meta);
+
+  const expression = o.importExpr(R3.declareNgModule).callFn([definitionMap.toLiteralMap()]);
+  const type = createNgModuleType(meta);
+
+  return {expression, type, additionalStatements: []};
+}
+
+function createNgModuleDefinitionMap(meta: R3NgModuleMetadata):
+    DefinitionMap<R3DeclareNgModuleMetadata> {
+  const definitionMap = new DefinitionMap<R3DeclareNgModuleMetadata>();
+
+  definitionMap.set('version', o.literal('0.0.0-PLACEHOLDER'));
+  definitionMap.set('ngImport', o.importExpr(R3.core));
+  definitionMap.set('type', meta.internalType);
+
+  // Only generate the keys in the metadata if the arrays have values.
+  // Note that for partial compilation we do not wrap forward-refs. This is handled in the linker.
+  // Hence passing `false` for `containsForwardDecls` in `refsToArray()` calls.
+  if (meta.bootstrap.length) {
+    definitionMap.set('bootstrap', refsToArray(meta.bootstrap, /* containsForwardDecls */ false));
+  }
+
+  if (meta.declarations.length) {
+    definitionMap.set(
+        'declarations', refsToArray(meta.declarations, /* containsForwardDecls */ false));
+  }
+
+  if (meta.imports.length) {
+    definitionMap.set('imports', refsToArray(meta.imports, /* containsForwardDecls */ false));
+  }
+
+  if (meta.exports.length) {
+    definitionMap.set('exports', refsToArray(meta.exports, /* containsForwardDecls */ false));
+  }
+
+  if (meta.schemas && meta.schemas.length) {
+    definitionMap.set('schemas', o.literalArr(meta.schemas.map(ref => ref.value)));
+  }
+
+  if (meta.containsForwardDecls) {
+    definitionMap.set('containsForwardDecls', o.literal(true));
+  }
+
+  if (meta.id) {
```

Comment:
```suggestion
  if (meta.id !== undefined) {
```
or could be `!== null`